### PR TITLE
Ipt service button fixes & cleanups

### PR DESCRIPTION
**Pull request by OzFalcon**
This is the first batch of IPT SERVICE button fixes & general minor button cleanups.
Hard coded button assignments are replace with standard inputs where possible.
The changes address issues listed here: http://mametesters.org/view.php?id=6489

### CODE REVIEW

File: src/mame/drivers/albazc.cpp [hunk 1]
```diff
@@ -207,12 +207,12 @@ static INPUT_PORTS_START( hanaroku )
 	PORT_BIT( 0x80, IP_ACTIVE_LOW, IPT_HANAFUDA_NO )
 
 	PORT_START("IN2")   /* 0xe002 */
-	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_SERVICE3 ) PORT_NAME("Data Clear")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_SERVICE4 ) PORT_NAME("Data Clear")
```
Comment (cuavas) [inline src/mame/drivers/albazc.cpp]: This is what `IPT_MEMORY_RESET` is for.

File: src/mame/drivers/albazc.cpp [hunk 2]
```diff
@@ -207,12 +207,12 @@ static INPUT_PORTS_START( hanaroku )
...
 	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_TILT )
-	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_BUTTON9 ) PORT_NAME("Medal In") PORT_CODE(KEYCODE_I)
-	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_BUTTON10 ) PORT_NAME("Pay Out") PORT_CODE(KEYCODE_O)
-	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_SERVICE1 ) PORT_NAME("Ext In 1")
-	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_SERVICE2 ) PORT_NAME("Ext In 2")
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_MAHJONG_I ) PORT_NAME("Medal In")
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_MAHJONG_O ) PORT_NAME("Pay Out")
```
Comment (cuavas) [inline src/mame/drivers/albazc.cpp]: These definitely shouldn't be I/O - medal in should probably be a coin or keyin, and payout should be `IPT_GAMBLE_KEYOUT` if it's an attendant keyswitch, or `IPT_GAMBLE_PAYOUT` if it's a player button.

Comment (OzFalcon) [inline src/mame/drivers/albazc.cpp]: They do not appear to be user controls. 
ie. Going into dip switch service mode - Card 1-6, Play, Start, Yes, No & D-Up - Will Register "PANEL OK" - So thats all the play requires.

I thought "Medal" could be referring to "Tokens", But credits appear to be covered with standard coin 1 & 2 inputs.

I'm not really sure how these ports are supposed to work.
  "Medal In" triggers "DIS" & "CLK" in the diagnostics screen.
  And "Pay Out" causes a "MEDAL EMPTY' message (Once > one credit is inserted).
Neither seem to alter credits (Perhaps it's waiting for a hopper response?).

I'd be inclined to set them as IPT_OTHER until their true nature is known.
Any suggestions on how to investigate this further?

Comment (OzFalcon) [inline src/mame/drivers/albazc.cpp]: I have also noticed the "Key" button highlights "Lockout" in the dip switch service mode.
This button should probably be a "PORT_TOGGLE" button.
ie. The Lockout "Key" is turned either on or off, Not Pushed and released.

Comment (OzFalcon) [inline src/mame/drivers/albazc.cpp]: I take that back (Using IPT_OTHER) - Your suggestion sounds more sensible.
I will change them to IPT_GAMBLE_KEYIN & IPT_GAMBLE_KEYOUT.
How they work - That's something else.

**Comment by galibert** [comment]
Sounds to me like the "pay out" bit is not a user-controllable line but
instead an empty signal that comes from the medal dispenser.  That's
usually implemented by having a read handler that reads the i/o port and
adds in the non-user bits.

Note that using user inputs on bit one does not understand to find out what
they do is a perfectly sane reverse-engineering strategy.

  OG.

**Comment by OzFalcon** [comment]
Ok, I have just worked out the "Medal In" and "Payout" function.
Once the "Payout" button is press, The game waits for a "Medal In" signal to indicate a Medal was dispensed.
Once it receives the signal - The game decreases the credit count.
(The payout dispenses only one medal per press)

So the "Payout" button should be changed to a "User" button (IPT_GAMBLE_PAYOUT).
And the "Medal In" button can be left as is (IPT_GAMBLE_KEYIN)

I will also make a small note somewhere in the driver about operation.

**Comment by cuavas** [comment]
OK, that kinda makes sense, but now we need to identify the output that it uses to dispense a medal and hook it up to hopper emulation…

**Comment by cuavas** [comment]
Also, you should probably understand the premise behind "medals".  They're coin-like tokens that you can buy at arcades to play these games, and you can win them from these games.  You can't redeem them for cash, only prizes or to play more games, although there's a black market involving exchanging them for cash.

The reason for medal games is that they're not regulated like gambling games.  In Japan, gambling games must be, at least technically, a game of skill (not chance), can't pay out jackpots over a certain level, are restricted to players over 18, and have various other rules applied.  As long as the medals can't officially be redeemed for cash, they don't count as gambling games, and the rules are far more lax.

**Comment by OzFalcon** [comment]
Yes, I read https://en.wikipedia.org/wiki/Medal_game 

I'm not sure about Mames "Key" ideology vs real world interfacing.
ie. Should an emulated key switch be a PORT_TOGGLE so it acts like a key switch.
Or should it be a button - So a real world key switch operates it correctly when connected.

**Comment by OzFalcon** [comment]
I think I have found the output for the hopper activation.
(My experience using the debugger is one step away from zero - I'm probably not using it right or efficiently)
From the main CPU program space memory I have eyeballed some address's.
I have labeled the address & values as best I can.

Activate Hopper: 0xC044
00 = Idle
01 = Dispense a Medal

Receive State: 0xC042
20 = Idle
60 = Expecting hopper signal

Payout Status: 0xC06A
00 = Idle
01 = Waiting for hopper signal
02 = Receiving hopper signal
03 = Error (Medal Empty or Medal Stick)

Plus there are some other addresses that deal with the timeouts.
I have no idea how to hook this up to the emulated hopper (I didn't even know Mame had an emulated Hopper!)

**Comment by OzFalcon** [comment]
Hookup hopper or popmessage has been added to the todo list.
Hopper activate addresses have been added to source notes.

**Comment by OzFalcon** [comment]
cuavas>>>> Also, have you checked for clashes after these changes?

Yes, I am checking for any problems like this.

**Comment by ajrhacker** [comment]
I figured out the hopper emulation for hanaroku, and added NVRAM while I was at it. This PR will need updating to avoid conflicts.

**Comment by OzFalcon** [comment]
I have updated the PR to coincide with your commit: https://github.com/mamedev/mame/commit/0bc96f38f1a2eb1e098882e474f9604ba905d972

File: src/mame/drivers/albazc.cpp (hunk 1 above)
Comment (OzFalcon) [inline src/mame/drivers/albazc.cpp]: IPT_MEMORY_RESET has been assign in commit: https://github.com/mamedev/mame/commit/0bc96f38f1a2eb1e098882e474f9604ba905d972

File: src/mame/drivers/albazc.cpp (hunk 2 above)
Comment (OzFalcon) [inline src/mame/drivers/albazc.cpp]: albazc.cpp PR has been updated to coincide with commit: https://github.com/mamedev/mame/commit/0bc96f38f1a2eb1e098882e474f9604ba905d972

**Pull request merged**